baybe/searchspace/discrete.py

```diff
@@ -230,6 +230,85 @@ def discrete_parameter_factory(
 
+    @classmethod
+    def from_simplex(
+        cls,
+        parameters: List[NumericalDiscreteParameter],
+        total: float,
+        boundary_only: bool = False,
+        tolerance: float = 1e-6,
+    ) -> SubspaceDiscrete:
+        """Efficiently create discrete simplex subspaces.
+
+        The same result can be achieved using
+        :meth:`baybe.searchspace.discrete.from_product` in combination with appropriate
+        sum constraints. However, such an approach is inefficient because the Cartesian
+        product involved creates an exponentially large set of candidates, most of
+        which do not satisfy the simplex constraints and must be subsequently be
+        filtered out by the method.
+
+        By contrast, this method uses a shortcut that removes invalid candidates
+        already during the creation of parameter combinations, resulting in a
+        significantly faster construction.
+
+        Args:
+            parameters: The parameter to be used for the simplex construction.
+            total: The desired sum of the parameter values defining the simplex size.
+            boundary_only: Flag determining whether to keep only parameter
+                configurations on the simplex boundary.
+            tolerance: Numerical tolerance used to validate the simplex constraint.
+
+        Raises:
+            ValueError: If the passed parameters are not suitable for a simplex
+                construction.
+
+        Returns:
+            The created simplex subspace.
+        """
+        # Validate parameter types
+        if not (all(isinstance(p, NumericalDiscreteParameter) for p in parameters)):
+            raise ValueError(
+                f"All parameters passed to '{cls.from_simplex.__name__}' "
+                f"must be of type '{NumericalDiscreteParameter.__name__}'."
+            )
+
+        # Validate non-negativity
+        min_values = [min(p.values) for p in parameters]
+        if not (min(min_values) >= 0.0):
+            raise ValueError(
+                f"All parameters passed to '{cls.from_simplex.__name__}' "
+                f"must have non-negative values only."
+            )
+
+        def drop_invalid(df: pd.DataFrame, equality: bool) -> None:
+            """Drop configuration that violate the simplex constraint."""
+            row_sums = df.sum(axis=1)
+            if equality:
+                rows_to_drop = row_sums[
+                    (row_sums < total - tolerance) | (row_sums > total + tolerance)
+                ].index
+            else:
+                rows_to_drop = row_sums[row_sums > total + tolerance].index
+            df.drop(rows_to_drop, inplace=True)
+
+        # Incrementally build up the space, dropping invalid configuration along the way
+        exp_rep = pd.DataFrame({parameters[0].name: parameters[0].values})
+        drop_invalid(exp_rep, equality=False)
+        for param in parameters[1:]:
```

Comment:
So what I get from here is that you iteratively add "the next" parameter and then filter out those tows whose sum now becomes too large, right?
This in particular means that (i) the performance of this method depends on the order of the parameters and (ii) that this might be inefficient, depending on the order of the parameters (imagine only one parameter being respondible for dropping rows, and that this happens to be the last).

The least I'd do is to put some information about this issue in the docstring of the function. Ideally, one could adjust the code such that these issues cannot occur at all. This should be in theory possible (or at least improvable) by e.g. re-ordering parameters or looking at specific algorithms for the SubsetSum problem.

HOWEVER, I know that this might not even be possible or a huge amount of additional work. I will thus also accept the current version but wanted to spark the discussion on that and hear your opinion :)

Reply:
True, the efficiency gains depend on the specific order of your parameter sequence. I am happy to add a note to the docstring mentioning this fact 👍🏼 But I also think (as you already said) that optimizing this part is a lot of work though there is (at least currently) no real use case that would benefit from it, so let's delay that until really needed. Nonetheless, appreciate your comment!

And BTW: I actually forgot to implement a slight improvement, and will do it now: instead of throwing out only configurations that exceed the sum at the current moment, we can additionally keep track of the minimum contribution that will come from the yet-to-be-added parameters and apply a stricter threshold that will be adjusted after each added parameter. Will ping you once it's there

Reply:
And the note is now there as well. Let me know if OK

Reply:
.... aaaaand the extension for arbitrary discrete parameter is not there as well. Done. Happy reviewing 🙃

Reply:
Looks good in general :) The parts that I have questions about are marked via comments, hence resolving this one.